src/main/java/com/ozonehis/ozone_demo_data/service/DemoDataService.java

```diff
@@ -0,0 +1,155 @@
+        ResponseEntity<String> response = restTemplate.postForEntity(generateDemoDataUrl, request, String.class);
+
+        validateResponse(response);
+        log.info("Demo data generation completed successfully");
+    }
+
+    HttpHeaders createAuthenticatedHeaders() {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+
+        String authToken = oauthEnabled ? obtainOAuthToken() : obtainBasicAuthToken();
+
+        if (oauthEnabled) {
+            headers.setBearerAuth(authToken);
+        } else {
+            headers.setBasicAuth(authToken);
+        }
+
+        return headers;
+    }
+
+    private String obtainOAuthToken() {
+        log.info("OAuth2 authentication enabled. Obtaining OAuth token...");
+        try (Keycloak keycloak = createKeycloakClient()) {
+            return keycloak.tokenManager().getAccessToken().getToken();
+        } catch (Exception e) {
+            throw new AuthenticationException("Failed to obtain OAuth token", e);
+        }
+    }
+
+    private Keycloak createKeycloakClient() {
+        return KeycloakBuilder.builder()
+                .serverUrl(keycloakConfig.getServerUrl())
+                .realm(KEYCLOAK_REALM)
+                .clientId(openmrsConfig.getClientId())
+                .clientSecret(openmrsConfig.getClientSecret())
+                .grantType("client_credentials")
+                .build();
+    }
+
+    private String obtainBasicAuthToken() {
+        log.info("Basic authentication enabled. Obtaining basic auth token...");
+        return Base64.getEncoder()
+                .encodeToString((openmrsConfig.getUsername() + ":" + openmrsConfig.getPassword())
+                        .getBytes(StandardCharsets.UTF_8));
+    }
+
+    Map<String, Object> createRequestBody() {
+        Map<String, Object> requestBody = new HashMap<>();
+        requestBody.put("numberOfDemoPatients", numberOfDemoPatients);
+        requestBody.put("createIfNotExists", true);
+        return requestBody;
+    }
+
+    String buildGenerateDemoDataUrl() {
+        return openmrsConfig.getUrl() + GENERATE_DEMO_DATA_ENDPOINT;
+    }
+
+    private void validateResponse(ResponseEntity<String> response) {
+        if (!response.getStatusCode().is2xxSuccessful()) {
+            throw new DemoDataGenerationException(
+                    "Demo data generation failed with status: " + response.getStatusCode());
+        }
+    }
+
+    private static class AuthenticationException extends RuntimeException {
+
+        public AuthenticationException(String message, Throwable cause) {
+            super(message, cause);
+        }
+    }
+
+    private static class DemoDataGenerationException extends RuntimeException {
+
+        public DemoDataGenerationException(String message) {
+            super(message);
+        }
+
+        public DemoDataGenerationException(String message, Throwable cause) {
+            super(message, cause);
+        }
+    }
+}
```

Comment:
Let's pull this class out into a different package

Reply:
I don't see the need as they are very specific to that `DemoDataService` but maybe for better organisation & readability.